lib/topics_filter.rb

```diff
@@ -495,29 +503,46 @@ def filter_users(values:)
             topics.id NOT IN (
               SELECT p1.topic_id
               FROM posts p1
-              WHERE p1.user_id IN (:user_ids) AND p1.deleted_at IS NULL
+              WHERE p1.user_id IN (:user_ids) AND p1.deleted_at IS NULL #{whisper_condition}
               GROUP BY p1.topic_id
               HAVING COUNT(DISTINCT p1.user_id) = :user_count
             )
           SQL
         else
-          user_ids.each_with_index { |uid, idx| @scope = @scope.where(<<~SQL) }
+          user_ids.each_with_index do |uid, idx|
+            whisper_condition_indexed =
+              if @guardian.can_see_whispers?
+                ""
+              else
+                "AND p#{idx}.post_type != #{Post.types[:whisper]}"
+              end
+
+            @scope = @scope.where(<<~SQL)
             EXISTS (
               SELECT 1
               FROM posts p#{idx}
-              WHERE p#{idx}.topic_id = topics.id AND p#{idx}.user_id = #{uid} AND p#{idx}.deleted_at IS NULL
+              WHERE p#{idx}.topic_id = topics.id AND p#{idx}.user_id = #{uid} AND p#{idx}.deleted_at IS NULL #{whisper_condition_indexed}
               LIMIT 1
             )
           SQL
+          end
         end
       else
         not_sql = prefix == "-" ? "NOT" : ""
+        whisper_condition_p =
+          if @guardian.can_see_whispers?
+            ""
+          else
+            "AND p.post_type != #{Post.types[:whisper]}"
+          end
```

Comment:
We are duplicating the logic here multiple times in the class. Maybe we can extract this into a method?

```
def whisper_clause(guardian, table_name)
  return "" if guardian.can_see_whispers?
  "AND #{table_name}.post_type != #{Post.types[:whisper]}"
end
```